Add ML-DSA post-quantum cryptography algorithm

HOCOOH opened the pull request:
为Tongsuo添加ML-DSA后量子签名算法支持，其中算法部分使用了CRYSTALS-DILITHIUM团队的官方参考实现。
文档PR链接：https://github.com/Tongsuo-Project/docs/pull/42

##### Checklist
<!-- 基于你的PR的实际情况移除不适用的项目。其他完成的项目修改[ ]为[x]. -->
- [x] 在 https://yuque.com/tsdoc 增加或更新了必要的文档
- [x] 增加或更新了必要的测试用例
- [ ] 对于重要修改，更新了CHANGES文件
- [ ] 当前修改存在对已有API参数或返回值的改变
- [ ] 当前修改存在对旧版本功能的兼容性改变（如网络协议或密码算法）

pr000000f commented on crypto/evp/evp_lib.c:
```diff
@@ -1180,6 +1182,15 @@ EVP_PKEY *EVP_PKEY_Q_keygen(OSSL_LIB_CTX *libctx, const char *propq,
+            ERR_raise(ERR_LIB_EVP, ERR_R_PASSED_INVALID_ARGUMENT);
+            goto end;
+        }
+        seed_len = va_arg(args, size_t);
```
Why do we need to explicitly provide a seed here? 
The other signatures only parse parameters such as the group name.
Also, in FIPS-204, Algorithm 1, the random seed is chosen inside KeyGen.

pr000000f commented on test/ml_dsa_internal_test.c:
```diff
@@ -0,0 +1,262 @@
+            goto err;
+        
+        params[0] = OSSL_PARAM_construct_octet_string(OSSL_SIGNATURE_PARAM_CONTEXT_STRING, ctx, sizeof(ctx));
+        params[1] = OSSL_PARAM_construct_int(OSSL_SIGNATURE_PARAM_DETERMINISTIC, &deterministic);
```
Is it possible to test the deterministic mode with a series of fixed test vectors including fixed key-pairs, messages and signatures?

pr000000f commented:
Hi, @HOCOOH, thank you for your contribution!
Could you please provide some manuals on how to activate the ML-DSA feature in configuration and on how to use the corresponding APIs? Maybe in a parallel PR for https://github.com/Tongsuo-Project/docs

HOCOOH committed:
Add ML-DSA post-quantum cryptography algorithm.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/Tongsuo-Project/Tongsuo?pullRequest=742) <br/>All committers have signed the CLA.

HOCOOH commented:
> Hi, @HOCOOH, thank you for your contribution! Could you please provide some manuals on how to activate the ML-DSA feature in configuration and on how to use the corresponding APIs? Maybe in a parallel PR for https://github.com/Tongsuo-Project/docs

Hi, @pr000000f 
Thanks for your review and comments! I have addressed the points you raised in the latest commit: 
1. The explicit `seed` parameter has been removed from the `EVP_PKEY_Q_keygen` API. The seed is now generated internally during key generation. An interface for setting the seed is preserved exclusively for testing purposes.
3. Added KAT tests for key generation and signature. These tests utilize deterministic inputs to verify that the outputs  match the expected results.
4. Fixed a few bugs related to the testing and EVP parameter setting.

The corresponding manual update has also been submitted in [PR #42](https://github.com/Tongsuo-Project/docs/pull/42) to the Tongsuo-Project/docs repository.

pr000000f commented:
> > Hi, @HOCOOH, thank you for your contribution! Could you please provide some manuals on how to activate the ML-DSA feature in configuration and on how to use the corresponding APIs? Maybe in a parallel PR for https://github.com/Tongsuo-Project/docs
> 
> Hi, @pr000000f Thanks for your review and comments! I have addressed the points you raised in the latest commit:
> 
> 1. The explicit `seed` parameter has been removed from the `EVP_PKEY_Q_keygen` API. The seed is now generated internally during key generation. An interface for setting the seed is preserved exclusively for testing purposes.
> 2. Added KAT tests for key generation and signature. These tests utilize deterministic inputs to verify that the outputs  match the expected results.
> 3. Fixed a few bugs related to the testing and EVP parameter setting.
> 
> The corresponding manual update has also been submitted in [PR #42](https://github.com/Tongsuo-Project/docs/pull/42) to the Tongsuo-Project/docs repository.

Thank you for the fix @HOCOOH!
Another question: How can I export a public key and a signature (since I want to verify its compatibility with OpenSSL 3.5.0)?  I tried PEM_write_bio_PUBKEY with a EVP_PKEY * but it failed.

athank2222 commented:
合并了算法，但不能签发证书，缺encoder/decoder:
root@xxx:/opt/tongsuopqc/bin# ./openssl genpkey -algorithm ml-dsa-65 -out ml-dsa-ca.key
Error writing key
80DB9927B17A0000:error:1D800065:ENCODER routines:OSSL_ENCODER_to_bio:reason(101):crypto/encode_decode/encoder_lib.c:55:No encoders were found. For standard encoders you need at least one of the default or base providers available. Did you forget to load them?
80DB9927B17A0000:error:04800073:PEM routines:do_pk8pkey:error converting private key:crypto/pem/pem_pk8.c:133:

HOCOOH committed:
Add ML-DSA codecs

HOCOOH commented:
@pr000000f @athank2222 
Thank you both for the review and for pointing this out. You were correct, the issue was the missing encoders/decoders.
I've pushed a new commit that implements them. This should resolve the errors you encountered.

Thanks again for your guidance

athank2222 commented:
After merge the encoder/decoder commit patch,  tried sign certificate commands, The following errors were found：
root@gcl:/opt/sm2tongsuo/bin# ./openssl genpkey -algorithm  ML-DSA-65 -out ca_mldsa65.key
root@gcl:/opt/sm2tongsuo/bin# ./openssl req -new -key ./ca_mldsa65.key -out ca_mldsa65.csr -subj "/C=CN/ST=Beijing/O=MyOrg/CN=MyCA"
Could not read private key from ./ca_mldsa65.key
4027790B997B0000:error:1608010C:STORE routines:ossl_store_handle_load_result:unsupported:crypto/store/store_result.c:151:
root@gcl:/opt/sm2tongsuo/bin# ./openssl list -decoders
Provided DECODERs:
  { 1.2.840.113549.1.1.1, 2.5.8.1.1, RSA, rsaEncryption } @ default (provider=default,fips=yes,input=der,structure=PrivateKeyInfo)
  { 1.2.840.113549.1.1.1, 2.5.8.1.1, RSA, rsaEncryption } @ default (provider=default,fips=yes,input=der,structure=SubjectPublicKeyInfo)
  { 1.2.840.113549.1.1.1, 2.5.8.1.1, RSA, rsaEncryption } @ default (provider=default,fips=yes,input=der,structure=type-specific)
  { 1.2.840.113549.1.1.1, 2.5.8.1.1, RSA, rsaEncryption } @ default (provider=default,fips=yes,input=der,structure=rsa)
  { 1.2.840.113549.1.1.1, 2.5.8.1.1, RSA, rsaEncryption } @ default (provider=default,fips=yes,input=msblob)
  { 1.2.840.113549.1.1.1, 2.5.8.1.1, RSA, rsaEncryption } @ default (provider=default,fips=yes,input=pvk)
  { 1.2.840.113549.1.3.1, DH, dhKeyAgreement } @ default (provider=default,fips=yes,input=der,structure=PrivateKeyInfo)
  { 1.2.840.113549.1.3.1, DH, dhKeyAgreement } @ default (provider=default,fips=yes,input=der,structure=SubjectPublicKeyInfo)
  { 1.2.840.113549.1.3.1, DH, dhKeyAgreement } @ default (provider=default,fips=yes,input=der,structure=type-specific)
  { 1.2.840.113549.1.3.1, DH, dhKeyAgreement } @ default (provider=default,fips=yes,input=der,structure=dh)
  { 1.2.840.10040.4.1, 1.2.840.10040.4.3, 1.3.14.3.2.12, 1.3.14.3.2.13, 1.3.14.3.2.27, DSA, DSA-old, DSA-SHA, DSA-SHA1, DSA-SHA1-old, dsaEncryption, dsaEncryption-old, dsaWithSHA, dsaWithSHA1, dsaWithSHA1-old } @ default (provider=default,fips=yes,input=der,structure=PrivateKeyInfo)
  { 1.2.840.10040.4.1, 1.2.840.10040.4.3, 1.3.14.3.2.12, 1.3.14.3.2.13, 1.3.14.3.2.27, DSA, DSA-old, DSA-SHA, DSA-SHA1, DSA-SHA1-old, dsaEncryption, dsaEncryption-old, dsaWithSHA, dsaWithSHA1, dsaWithSHA1-old } @ default (provider=default,fips=yes,input=der,structure=SubjectPublicKeyInfo)
  { 1.2.840.10040.4.1, 1.2.840.10040.4.3, 1.3.14.3.2.12, 1.3.14.3.2.13, 1.3.14.3.2.27, DSA, DSA-old, DSA-SHA, DSA-SHA1, DSA-SHA1-old, dsaEncryption, dsaEncryption-old, dsaWithSHA, dsaWithSHA1, dsaWithSHA1-old } @ default (provider=default,fips=yes,input=der,structure=type-specific)
  { 1.2.840.10040.4.1, 1.2.840.10040.4.3, 1.3.14.3.2.12, 1.3.14.3.2.13, 1.3.14.3.2.27, DSA, DSA-old, DSA-SHA, DSA-SHA1, DSA-SHA1-old, dsaEncryption, dsaEncryption-old, dsaWithSHA, dsaWithSHA1, dsaWithSHA1-old } @ default (provider=default,fips=yes,input=der,structure=dsa)
  { 1.2.840.10040.4.1, 1.2.840.10040.4.3, 1.3.14.3.2.12, 1.3.14.3.2.13, 1.3.14.3.2.27, DSA, DSA-old, DSA-SHA, DSA-SHA1, DSA-SHA1-old, dsaEncryption, dsaEncryption-old, dsaWithSHA, dsaWithSHA1, dsaWithSHA1-old } @ default (provider=default,fips=yes,input=msblob)
  { 1.2.840.10040.4.1, 1.2.840.10040.4.3, 1.3.14.3.2.12, 1.3.14.3.2.13, 1.3.14.3.2.27, DSA, DSA-old, DSA-SHA, DSA-SHA1, DSA-SHA1-old, dsaEncryption, dsaEncryption-old, dsaWithSHA, dsaWithSHA1, dsaWithSHA1-old } @ default (provider=default,fips=yes,input=pvk)
  { 1.2.840.10045.2.1, EC, id-ecPublicKey } @ default (provider=default,fips=yes,input=der,structure=PrivateKeyInfo)
  { 1.2.840.10045.2.1, EC, id-ecPublicKey } @ default (provider=default,fips=yes,input=der,structure=SubjectPublicKeyInfo)
  { 1.2.840.10045.2.1, EC, id-ecPublicKey } @ default (provider=default,fips=yes,input=der,structure=type-specific)
  { 1.2.840.10045.2.1, EC, id-ecPublicKey } @ default (provider=default,fips=yes,input=der,structure=ec)
  { 1.2.840.113549.1.1.10, RSA-PSS, RSASSA-PSS, rsassaPss } @ default (provider=default,fips=yes,input=der,structure=PrivateKeyInfo)
  { 1.2.840.113549.1.1.10, RSA-PSS, RSASSA-PSS, rsassaPss } @ default (provider=default,fips=yes,input=der,structure=SubjectPublicKeyInfo)
  { 1.2.840.10046.2.1, dhpublicnumber, DHX, X9.42 DH } @ default (provider=default,fips=yes,input=der,structure=PrivateKeyInfo)
  { 1.2.840.10046.2.1, dhpublicnumber, DHX, X9.42 DH } @ default (provider=default,fips=yes,input=der,structure=SubjectPublicKeyInfo)
  { 1.2.840.10046.2.1, dhpublicnumber, DHX, X9.42 DH } @ default (provider=default,fips=yes,input=der,structure=type-specific)
  { 1.2.840.10046.2.1, dhpublicnumber, DHX, X9.42 DH } @ default (provider=default,fips=yes,input=der,structure=dhx)
  { 1.3.101.110, X25519 } @ default (provider=default,fips=yes,input=der,structure=PrivateKeyInfo)
  { 1.3.101.110, X25519 } @ default (provider=default,fips=yes,input=der,structure=SubjectPublicKeyInfo)
  { 1.3.101.111, X448 } @ default (provider=default,fips=yes,input=der,structure=PrivateKeyInfo)
  { 1.3.101.111, X448 } @ default (provider=default,fips=yes,input=der,structure=SubjectPublicKeyInfo)
  { 1.3.101.112, ED25519 } @ default (provider=default,fips=yes,input=der,structure=PrivateKeyInfo)
  { 1.3.101.112, ED25519 } @ default (provider=default,fips=yes,input=der,structure=SubjectPublicKeyInfo)
  { 1.3.101.113, ED448 } @ default (provider=default,fips=yes,input=der,structure=PrivateKeyInfo)
  { 1.3.101.113, ED448 } @ default (provider=default,fips=yes,input=der,structure=SubjectPublicKeyInfo)
  { 1.2.156.10197.1.301, SM2 } @ default (provider=default,fips=no,input=der,structure=PrivateKeyInfo)
  { 1.2.156.10197.1.301, SM2 } @ default (provider=default,fips=no,input=der,structure=SubjectPublicKeyInfo)
  { 1.2.156.10197.1.301, SM2 } @ default (provider=default,fips=no,input=der,structure=type-specific)
  ML-DSA-65 @ default (provider=default,fips=no,input=der,structure=PrivateKeyInfo)
  ML-DSA-65 @ default (provider=default,fips=no,input=der,structure=SubjectPublicKeyInfo)
  DER @ default (provider=default,fips=yes,input=der,structure=SubjectPublicKeyInfo)
  DER @ default (provider=default,fips=yes,input=pem)
  DER @ default (provider=default,fips=yes,input=der,structure=EncryptedPrivateKeyInfo)
root@gcl:/opt/sm2tongsuo/bin#

HOCOOH commented:
> After merge the encoder/decoder commit patch, tried sign certificate commands, The following errors were found： root@gcl:/opt/sm2tongsuo/bin# ./openssl genpkey -algorithm ML-DSA-65 -out ca_mldsa65.key root@gcl:/opt/sm2tongsuo/bin# ./openssl req -new -key ./ca_mldsa65.key -out ca_mldsa65.csr -subj "/C=CN/ST=Beijing/O=MyOrg/CN=MyCA" Could not read private key from ./ca_mldsa65.key 4027790B997B0000:error:1608010C:STORE routines:ossl_store_handle_load_result:unsupported:crypto/store/store_result.c:151: root@gcl:/opt/sm2tongsuo/bin# ./openssl list -decoders Provided DECODERs: { 1.2.840.113549.1.1.1, 2.5.8.1.1, RSA, rsaEncryption } @ default (provider=default,fips=yes,input=der,structure=PrivateKeyInfo) { 1.2.840.113549.1.1.1, 2.5.8.1.1, RSA, rsaEncryption } @ default (provider=default,fips=yes,input=der,structure=SubjectPublicKeyInfo) { 1.2.840.113549.1.1.1, 2.5.8.1.1, RSA, rsaEncryption } @ default (provider=default,fips=yes,input=der,structure=type-specific) { 1.2.840.113549.1.1.1, 2.5.8.1.1, RSA, rsaEncryption } @ default (provider=default,fips=yes,input=der,structure=rsa) { 1.2.840.113549.1.1.1, 2.5.8.1.1, RSA, rsaEncryption } @ default (provider=default,fips=yes,input=msblob) { 1.2.840.113549.1.1.1, 2.5.8.1.1, RSA, rsaEncryption } @ default (provider=default,fips=yes,input=pvk) { 1.2.840.113549.1.3.1, DH, dhKeyAgreement } @ default (provider=default,fips=yes,input=der,structure=PrivateKeyInfo) { 1.2.840.113549.1.3.1, DH, dhKeyAgreement } @ default (provider=default,fips=yes,input=der,structure=SubjectPublicKeyInfo) { 1.2.840.113549.1.3.1, DH, dhKeyAgreement } @ default (provider=default,fips=yes,input=der,structure=type-specific) { 1.2.840.113549.1.3.1, DH, dhKeyAgreement } @ default (provider=default,fips=yes,input=der,structure=dh) { 1.2.840.10040.4.1, 1.2.840.10040.4.3, 1.3.14.3.2.12, 1.3.14.3.2.13, 1.3.14.3.2.27, DSA, DSA-old, DSA-SHA, DSA-SHA1, DSA-SHA1-old, dsaEncryption, dsaEncryption-old, dsaWithSHA, dsaWithSHA1, dsaWithSHA1-old } @ default (provider=default,fips=yes,input=der,structure=PrivateKeyInfo) { 1.2.840.10040.4.1, 1.2.840.10040.4.3, 1.3.14.3.2.12, 1.3.14.3.2.13, 1.3.14.3.2.27, DSA, DSA-old, DSA-SHA, DSA-SHA1, DSA-SHA1-old, dsaEncryption, dsaEncryption-old, dsaWithSHA, dsaWithSHA1, dsaWithSHA1-old } @ default (provider=default,fips=yes,input=der,structure=SubjectPublicKeyInfo) { 1.2.840.10040.4.1, 1.2.840.10040.4.3, 1.3.14.3.2.12, 1.3.14.3.2.13, 1.3.14.3.2.27, DSA, DSA-old, DSA-SHA, DSA-SHA1, DSA-SHA1-old, dsaEncryption, dsaEncryption-old, dsaWithSHA, dsaWithSHA1, dsaWithSHA1-old } @ default (provider=default,fips=yes,input=der,structure=type-specific) { 1.2.840.10040.4.1, 1.2.840.10040.4.3, 1.3.14.3.2.12, 1.3.14.3.2.13, 1.3.14.3.2.27, DSA, DSA-old, DSA-SHA, DSA-SHA1, DSA-SHA1-old, dsaEncryption, dsaEncryption-old, dsaWithSHA, dsaWithSHA1, dsaWithSHA1-old } @ default (provider=default,fips=yes,input=der,structure=dsa) { 1.2.840.10040.4.1, 1.2.840.10040.4.3, 1.3.14.3.2.12, 1.3.14.3.2.13, 1.3.14.3.2.27, DSA, DSA-old, DSA-SHA, DSA-SHA1, DSA-SHA1-old, dsaEncryption, dsaEncryption-old, dsaWithSHA, dsaWithSHA1, dsaWithSHA1-old } @ default (provider=default,fips=yes,input=msblob) { 1.2.840.10040.4.1, 1.2.840.10040.4.3, 1.3.14.3.2.12, 1.3.14.3.2.13, 1.3.14.3.2.27, DSA, DSA-old, DSA-SHA, DSA-SHA1, DSA-SHA1-old, dsaEncryption, dsaEncryption-old, dsaWithSHA, dsaWithSHA1, dsaWithSHA1-old } @ default (provider=default,fips=yes,input=pvk) { 1.2.840.10045.2.1, EC, id-ecPublicKey } @ default (provider=default,fips=yes,input=der,structure=PrivateKeyInfo) { 1.2.840.10045.2.1, EC, id-ecPublicKey } @ default (provider=default,fips=yes,input=der,structure=SubjectPublicKeyInfo) { 1.2.840.10045.2.1, EC, id-ecPublicKey } @ default (provider=default,fips=yes,input=der,structure=type-specific) { 1.2.840.10045.2.1, EC, id-ecPublicKey } @ default (provider=default,fips=yes,input=der,structure=ec) { 1.2.840.113549.1.1.10, RSA-PSS, RSASSA-PSS, rsassaPss } @ default (provider=default,fips=yes,input=der,structure=PrivateKeyInfo) { 1.2.840.113549.1.1.10, RSA-PSS, RSASSA-PSS, rsassaPss } @ default (provider=default,fips=yes,input=der,structure=SubjectPublicKeyInfo) { 1.2.840.10046.2.1, dhpublicnumber, DHX, X9.42 DH } @ default (provider=default,fips=yes,input=der,structure=PrivateKeyInfo) { 1.2.840.10046.2.1, dhpublicnumber, DHX, X9.42 DH } @ default (provider=default,fips=yes,input=der,structure=SubjectPublicKeyInfo) { 1.2.840.10046.2.1, dhpublicnumber, DHX, X9.42 DH } @ default (provider=default,fips=yes,input=der,structure=type-specific) { 1.2.840.10046.2.1, dhpublicnumber, DHX, X9.42 DH } @ default (provider=default,fips=yes,input=der,structure=dhx) { 1.3.101.110, X25519 } @ default (provider=default,fips=yes,input=der,structure=PrivateKeyInfo) { 1.3.101.110, X25519 } @ default (provider=default,fips=yes,input=der,structure=SubjectPublicKeyInfo) { 1.3.101.111, X448 } @ default (provider=default,fips=yes,input=der,structure=PrivateKeyInfo) { 1.3.101.111, X448 } @ default (provider=default,fips=yes,input=der,structure=SubjectPublicKeyInfo) { 1.3.101.112, ED25519 } @ default (provider=default,fips=yes,input=der,structure=PrivateKeyInfo) { 1.3.101.112, ED25519 } @ default (provider=default,fips=yes,input=der,structure=SubjectPublicKeyInfo) { 1.3.101.113, ED448 } @ default (provider=default,fips=yes,input=der,structure=PrivateKeyInfo) { 1.3.101.113, ED448 } @ default (provider=default,fips=yes,input=der,structure=SubjectPublicKeyInfo) { 1.2.156.10197.1.301, SM2 } @ default (provider=default,fips=no,input=der,structure=PrivateKeyInfo) { 1.2.156.10197.1.301, SM2 } @ default (provider=default,fips=no,input=der,structure=SubjectPublicKeyInfo) { 1.2.156.10197.1.301, SM2 } @ default (provider=default,fips=no,input=der,structure=type-specific) ML-DSA-65 @ default (provider=default,fips=no,input=der,structure=PrivateKeyInfo) ML-DSA-65 @ default (provider=default,fips=no,input=der,structure=SubjectPublicKeyInfo) DER @ default (provider=default,fips=yes,input=der,structure=SubjectPublicKeyInfo) DER @ default (provider=default,fips=yes,input=pem) DER @ default (provider=default,fips=yes,input=der,structure=EncryptedPrivateKeyInfo) root@gcl:/opt/sm2tongsuo/bin#

Thanks for the feedback. This seems like an environment-specific issue, as I cannot reproduce it locally. Could you please share some details about your build environment so I can try to replicate it? Specifically:

1. OS and architecture.
2. The commit hash.
3. The full ./config command you used.
4. The output of gcc --version.

Thanks for your help!

athank2222 commented:
> > After merge the encoder/decoder commit patch, tried sign certificate commands, The following errors were found： root@gcl:/opt/sm2tongsuo/bin# ./openssl genpkey -algorithm ML-DSA-65 -out ca_mldsa65.key root@gcl:/opt/sm2tongsuo/bin# ./openssl req -new -key ./ca_mldsa65.key -out ca_mldsa65.csr -subj "/C=CN/ST=Beijing/O=MyOrg/CN=MyCA" Could not read private key from ./ca_mldsa65.key 4027790B997B0000:error:1608010C:STORE routines:ossl_store_handle_load_result:unsupported:crypto/store/store_result.c:151: root@gcl:/opt/sm2tongsuo/bin# ./openssl list -decoders Provided DECODERs: { 1.2.840.113549.1.1.1, 2.5.8.1.1, RSA, rsaEncryption } @ default (provider=default,fips=yes,input=der,structure=PrivateKeyInfo) { 1.2.840.113549.1.1.1, 2.5.8.1.1, RSA, rsaEncryption } @ default (provider=default,fips=yes,input=der,structure=SubjectPublicKeyInfo) { 1.2.840.113549.1.1.1, 2.5.8.1.1, RSA, rsaEncryption } @ default (provider=default,fips=yes,input=der,structure=type-specific) { 1.2.840.113549.1.1.1, 2.5.8.1.1, RSA, rsaEncryption } @ default (provider=default,fips=yes,input=der,structure=rsa) { 1.2.840.113549.1.1.1, 2.5.8.1.1, RSA, rsaEncryption } @ default (provider=default,fips=yes,input=msblob) { 1.2.840.113549.1.1.1, 2.5.8.1.1, RSA, rsaEncryption } @ default (provider=default,fips=yes,input=pvk) { 1.2.840.113549.1.3.1, DH, dhKeyAgreement } @ default (provider=default,fips=yes,input=der,structure=PrivateKeyInfo) { 1.2.840.113549.1.3.1, DH, dhKeyAgreement } @ default (provider=default,fips=yes,input=der,structure=SubjectPublicKeyInfo) { 1.2.840.113549.1.3.1, DH, dhKeyAgreement } @ default (provider=default,fips=yes,input=der,structure=type-specific) { 1.2.840.113549.1.3.1, DH, dhKeyAgreement } @ default (provider=default,fips=yes,input=der,structure=dh) { 1.2.840.10040.4.1, 1.2.840.10040.4.3, 1.3.14.3.2.12, 1.3.14.3.2.13, 1.3.14.3.2.27, DSA, DSA-old, DSA-SHA, DSA-SHA1, DSA-SHA1-old, dsaEncryption, dsaEncryption-old, dsaWithSHA, dsaWithSHA1, dsaWithSHA1-old } @ default (provider=default,fips=yes,input=der,structure=PrivateKeyInfo) { 1.2.840.10040.4.1, 1.2.840.10040.4.3, 1.3.14.3.2.12, 1.3.14.3.2.13, 1.3.14.3.2.27, DSA, DSA-old, DSA-SHA, DSA-SHA1, DSA-SHA1-old, dsaEncryption, dsaEncryption-old, dsaWithSHA, dsaWithSHA1, dsaWithSHA1-old } @ default (provider=default,fips=yes,input=der,structure=SubjectPublicKeyInfo) { 1.2.840.10040.4.1, 1.2.840.10040.4.3, 1.3.14.3.2.12, 1.3.14.3.2.13, 1.3.14.3.2.27, DSA, DSA-old, DSA-SHA, DSA-SHA1, DSA-SHA1-old, dsaEncryption, dsaEncryption-old, dsaWithSHA, dsaWithSHA1, dsaWithSHA1-old } @ default (provider=default,fips=yes,input=der,structure=type-specific) { 1.2.840.10040.4.1, 1.2.840.10040.4.3, 1.3.14.3.2.12, 1.3.14.3.2.13, 1.3.14.3.2.27, DSA, DSA-old, DSA-SHA, DSA-SHA1, DSA-SHA1-old, dsaEncryption, dsaEncryption-old, dsaWithSHA, dsaWithSHA1, dsaWithSHA1-old } @ default (provider=default,fips=yes,input=der,structure=dsa) { 1.2.840.10040.4.1, 1.2.840.10040.4.3, 1.3.14.3.2.12, 1.3.14.3.2.13, 1.3.14.3.2.27, DSA, DSA-old, DSA-SHA, DSA-SHA1, DSA-SHA1-old, dsaEncryption, dsaEncryption-old, dsaWithSHA, dsaWithSHA1, dsaWithSHA1-old } @ default (provider=default,fips=yes,input=msblob) { 1.2.840.10040.4.1, 1.2.840.10040.4.3, 1.3.14.3.2.12, 1.3.14.3.2.13, 1.3.14.3.2.27, DSA, DSA-old, DSA-SHA, DSA-SHA1, DSA-SHA1-old, dsaEncryption, dsaEncryption-old, dsaWithSHA, dsaWithSHA1, dsaWithSHA1-old } @ default (provider=default,fips=yes,input=pvk) { 1.2.840.10045.2.1, EC, id-ecPublicKey } @ default (provider=default,fips=yes,input=der,structure=PrivateKeyInfo) { 1.2.840.10045.2.1, EC, id-ecPublicKey } @ default (provider=default,fips=yes,input=der,structure=SubjectPublicKeyInfo) { 1.2.840.10045.2.1, EC, id-ecPublicKey } @ default (provider=default,fips=yes,input=der,structure=type-specific) { 1.2.840.10045.2.1, EC, id-ecPublicKey } @ default (provider=default,fips=yes,input=der,structure=ec) { 1.2.840.113549.1.1.10, RSA-PSS, RSASSA-PSS, rsassaPss } @ default (provider=default,fips=yes,input=der,structure=PrivateKeyInfo) { 1.2.840.113549.1.1.10, RSA-PSS, RSASSA-PSS, rsassaPss } @ default (provider=default,fips=yes,input=der,structure=SubjectPublicKeyInfo) { 1.2.840.10046.2.1, dhpublicnumber, DHX, X9.42 DH } @ default (provider=default,fips=yes,input=der,structure=PrivateKeyInfo) { 1.2.840.10046.2.1, dhpublicnumber, DHX, X9.42 DH } @ default (provider=default,fips=yes,input=der,structure=SubjectPublicKeyInfo) { 1.2.840.10046.2.1, dhpublicnumber, DHX, X9.42 DH } @ default (provider=default,fips=yes,input=der,structure=type-specific) { 1.2.840.10046.2.1, dhpublicnumber, DHX, X9.42 DH } @ default (provider=default,fips=yes,input=der,structure=dhx) { 1.3.101.110, X25519 } @ default (provider=default,fips=yes,input=der,structure=PrivateKeyInfo) { 1.3.101.110, X25519 } @ default (provider=default,fips=yes,input=der,structure=SubjectPublicKeyInfo) { 1.3.101.111, X448 } @ default (provider=default,fips=yes,input=der,structure=PrivateKeyInfo) { 1.3.101.111, X448 } @ default (provider=default,fips=yes,input=der,structure=SubjectPublicKeyInfo) { 1.3.101.112, ED25519 } @ default (provider=default,fips=yes,input=der,structure=PrivateKeyInfo) { 1.3.101.112, ED25519 } @ default (provider=default,fips=yes,input=der,structure=SubjectPublicKeyInfo) { 1.3.101.113, ED448 } @ default (provider=default,fips=yes,input=der,structure=PrivateKeyInfo) { 1.3.101.113, ED448 } @ default (provider=default,fips=yes,input=der,structure=SubjectPublicKeyInfo) { 1.2.156.10197.1.301, SM2 } @ default (provider=default,fips=no,input=der,structure=PrivateKeyInfo) { 1.2.156.10197.1.301, SM2 } @ default (provider=default,fips=no,input=der,structure=SubjectPublicKeyInfo) { 1.2.156.10197.1.301, SM2 } @ default (provider=default,fips=no,input=der,structure=type-specific) ML-DSA-65 @ default (provider=default,fips=no,input=der,structure=PrivateKeyInfo) ML-DSA-65 @ default (provider=default,fips=no,input=der,structure=SubjectPublicKeyInfo) DER @ default (provider=default,fips=yes,input=der,structure=SubjectPublicKeyInfo) DER @ default (provider=default,fips=yes,input=pem) DER @ default (provider=default,fips=yes,input=der,structure=EncryptedPrivateKeyInfo) root@gcl:/opt/sm2tongsuo/bin#
> 
> Thanks for the feedback. This seems like an environment-specific issue, as I cannot reproduce it locally. Could you please share some details about your build environment so I can try to replicate it? Specifically:
> 
> 1. OS and architecture.
> 2. The commit hash.
> 3. The full ./config command you used.
> 4. The output of gcc --version.
> 
> Thanks for your help!

请忽略上面的问题，是代码合并引起的问题。

athank2222 commented:
> > After merge the encoder/decoder commit patch, tried sign certificate commands, The following errors were found： root@gcl:/opt/sm2tongsuo/bin# ./openssl genpkey -algorithm ML-DSA-65 -out ca_mldsa65.key root@gcl:/opt/sm2tongsuo/bin# ./openssl req -new -key ./ca_mldsa65.key -out ca_mldsa65.csr -subj "/C=CN/ST=Beijing/O=MyOrg/CN=MyCA" Could not read private key from ./ca_mldsa65.key 4027790B997B0000:error:1608010C:STORE routines:ossl_store_handle_load_result:unsupported:crypto/store/store_result.c:151: root@gcl:/opt/sm2tongsuo/bin# ./openssl list -decoders Provided DECODERs: { 1.2.840.113549.1.1.1, 2.5.8.1.1, RSA, rsaEncryption } @ default (provider=default,fips=yes,input=der,structure=PrivateKeyInfo) { 1.2.840.113549.1.1.1, 2.5.8.1.1, RSA, rsaEncryption } @ default (provider=default,fips=yes,input=der,structure=SubjectPublicKeyInfo) { 1.2.840.113549.1.1.1, 2.5.8.1.1, RSA, rsaEncryption } @ default (provider=default,fips=yes,input=der,structure=type-specific) { 1.2.840.113549.1.1.1, 2.5.8.1.1, RSA, rsaEncryption } @ default (provider=default,fips=yes,input=der,structure=rsa) { 1.2.840.113549.1.1.1, 2.5.8.1.1, RSA, rsaEncryption } @ default (provider=default,fips=yes,input=msblob) { 1.2.840.113549.1.1.1, 2.5.8.1.1, RSA, rsaEncryption } @ default (provider=default,fips=yes,input=pvk) { 1.2.840.113549.1.3.1, DH, dhKeyAgreement } @ default (provider=default,fips=yes,input=der,structure=PrivateKeyInfo) { 1.2.840.113549.1.3.1, DH, dhKeyAgreement } @ default (provider=default,fips=yes,input=der,structure=SubjectPublicKeyInfo) { 1.2.840.113549.1.3.1, DH, dhKeyAgreement } @ default (provider=default,fips=yes,input=der,structure=type-specific) { 1.2.840.113549.1.3.1, DH, dhKeyAgreement } @ default (provider=default,fips=yes,input=der,structure=dh) { 1.2.840.10040.4.1, 1.2.840.10040.4.3, 1.3.14.3.2.12, 1.3.14.3.2.13, 1.3.14.3.2.27, DSA, DSA-old, DSA-SHA, DSA-SHA1, DSA-SHA1-old, dsaEncryption, dsaEncryption-old, dsaWithSHA, dsaWithSHA1, dsaWithSHA1-old } @ default (provider=default,fips=yes,input=der,structure=PrivateKeyInfo) { 1.2.840.10040.4.1, 1.2.840.10040.4.3, 1.3.14.3.2.12, 1.3.14.3.2.13, 1.3.14.3.2.27, DSA, DSA-old, DSA-SHA, DSA-SHA1, DSA-SHA1-old, dsaEncryption, dsaEncryption-old, dsaWithSHA, dsaWithSHA1, dsaWithSHA1-old } @ default (provider=default,fips=yes,input=der,structure=SubjectPublicKeyInfo) { 1.2.840.10040.4.1, 1.2.840.10040.4.3, 1.3.14.3.2.12, 1.3.14.3.2.13, 1.3.14.3.2.27, DSA, DSA-old, DSA-SHA, DSA-SHA1, DSA-SHA1-old, dsaEncryption, dsaEncryption-old, dsaWithSHA, dsaWithSHA1, dsaWithSHA1-old } @ default (provider=default,fips=yes,input=der,structure=type-specific) { 1.2.840.10040.4.1, 1.2.840.10040.4.3, 1.3.14.3.2.12, 1.3.14.3.2.13, 1.3.14.3.2.27, DSA, DSA-old, DSA-SHA, DSA-SHA1, DSA-SHA1-old, dsaEncryption, dsaEncryption-old, dsaWithSHA, dsaWithSHA1, dsaWithSHA1-old } @ default (provider=default,fips=yes,input=der,structure=dsa) { 1.2.840.10040.4.1, 1.2.840.10040.4.3, 1.3.14.3.2.12, 1.3.14.3.2.13, 1.3.14.3.2.27, DSA, DSA-old, DSA-SHA, DSA-SHA1, DSA-SHA1-old, dsaEncryption, dsaEncryption-old, dsaWithSHA, dsaWithSHA1, dsaWithSHA1-old } @ default (provider=default,fips=yes,input=msblob) { 1.2.840.10040.4.1, 1.2.840.10040.4.3, 1.3.14.3.2.12, 1.3.14.3.2.13, 1.3.14.3.2.27, DSA, DSA-old, DSA-SHA, DSA-SHA1, DSA-SHA1-old, dsaEncryption, dsaEncryption-old, dsaWithSHA, dsaWithSHA1, dsaWithSHA1-old } @ default (provider=default,fips=yes,input=pvk) { 1.2.840.10045.2.1, EC, id-ecPublicKey } @ default (provider=default,fips=yes,input=der,structure=PrivateKeyInfo) { 1.2.840.10045.2.1, EC, id-ecPublicKey } @ default (provider=default,fips=yes,input=der,structure=SubjectPublicKeyInfo) { 1.2.840.10045.2.1, EC, id-ecPublicKey } @ default (provider=default,fips=yes,input=der,structure=type-specific) { 1.2.840.10045.2.1, EC, id-ecPublicKey } @ default (provider=default,fips=yes,input=der,structure=ec) { 1.2.840.113549.1.1.10, RSA-PSS, RSASSA-PSS, rsassaPss } @ default (provider=default,fips=yes,input=der,structure=PrivateKeyInfo) { 1.2.840.113549.1.1.10, RSA-PSS, RSASSA-PSS, rsassaPss } @ default (provider=default,fips=yes,input=der,structure=SubjectPublicKeyInfo) { 1.2.840.10046.2.1, dhpublicnumber, DHX, X9.42 DH } @ default (provider=default,fips=yes,input=der,structure=PrivateKeyInfo) { 1.2.840.10046.2.1, dhpublicnumber, DHX, X9.42 DH } @ default (provider=default,fips=yes,input=der,structure=SubjectPublicKeyInfo) { 1.2.840.10046.2.1, dhpublicnumber, DHX, X9.42 DH } @ default (provider=default,fips=yes,input=der,structure=type-specific) { 1.2.840.10046.2.1, dhpublicnumber, DHX, X9.42 DH } @ default (provider=default,fips=yes,input=der,structure=dhx) { 1.3.101.110, X25519 } @ default (provider=default,fips=yes,input=der,structure=PrivateKeyInfo) { 1.3.101.110, X25519 } @ default (provider=default,fips=yes,input=der,structure=SubjectPublicKeyInfo) { 1.3.101.111, X448 } @ default (provider=default,fips=yes,input=der,structure=PrivateKeyInfo) { 1.3.101.111, X448 } @ default (provider=default,fips=yes,input=der,structure=SubjectPublicKeyInfo) { 1.3.101.112, ED25519 } @ default (provider=default,fips=yes,input=der,structure=PrivateKeyInfo) { 1.3.101.112, ED25519 } @ default (provider=default,fips=yes,input=der,structure=SubjectPublicKeyInfo) { 1.3.101.113, ED448 } @ default (provider=default,fips=yes,input=der,structure=PrivateKeyInfo) { 1.3.101.113, ED448 } @ default (provider=default,fips=yes,input=der,structure=SubjectPublicKeyInfo) { 1.2.156.10197.1.301, SM2 } @ default (provider=default,fips=no,input=der,structure=PrivateKeyInfo) { 1.2.156.10197.1.301, SM2 } @ default (provider=default,fips=no,input=der,structure=SubjectPublicKeyInfo) { 1.2.156.10197.1.301, SM2 } @ default (provider=default,fips=no,input=der,structure=type-specific) ML-DSA-65 @ default (provider=default,fips=no,input=der,structure=PrivateKeyInfo) ML-DSA-65 @ default (provider=default,fips=no,input=der,structure=SubjectPublicKeyInfo) DER @ default (provider=default,fips=yes,input=der,structure=SubjectPublicKeyInfo) DER @ default (provider=default,fips=yes,input=pem) DER @ default (provider=default,fips=yes,input=der,structure=EncryptedPrivateKeyInfo) root@gcl:/opt/sm2tongsuo/bin#
> 
> Thanks for the feedback. This seems like an environment-specific issue, as I cannot reproduce it locally. Could you please share some details about your build environment so I can try to replicate it? Specifically:
> 
> 1. OS and architecture.
> 2. The commit hash.
> 3. The full ./config command you used.
> 4. The output of gcc --version.
> 
> Thanks for your help!

签发证书还需要在 obj_xref.txt 文件里加ML_DSA_65，文件内容
[obj_xref.txt](https://github.com/user-attachments/files/21545584/obj_xref.txt)
21  ED448		    undef	ED448
22 #ML_DSA_44  undef ML_DSA_44
23 ML_DSA_65  undef ML_DSA_65
24 #ML_DSA_87  undef ML_DSA_87

Maxlsc commented:
Hi, @HOCOOH, We checked the interoperability of your code with OpenSSL 3.5.2 using the API and command line for signature verification and signing certificate , and found no problems.

HOCOOH commented:
> > > After merge the encoder/decoder commit patch, tried sign certificate commands, The following errors were found： root@gcl:/opt/sm2tongsuo/bin# ./openssl genpkey -algorithm ML-DSA-65 -out ca_mldsa65.key root@gcl:/opt/sm2tongsuo/bin# ./openssl req -new -key ./ca_mldsa65.key -out ca_mldsa65.csr -subj "/C=CN/ST=Beijing/O=MyOrg/CN=MyCA" Could not read private key from ./ca_mldsa65.key 4027790B997B0000:error:1608010C:STORE routines:ossl_store_handle_load_result:unsupported:crypto/store/store_result.c:151: root@gcl:/opt/sm2tongsuo/bin# ./openssl list -decoders Provided DECODERs: { 1.2.840.113549.1.1.1, 2.5.8.1.1, RSA, rsaEncryption } @ default (provider=default,fips=yes,input=der,structure=PrivateKeyInfo) { 1.2.840.113549.1.1.1, 2.5.8.1.1, RSA, rsaEncryption } @ default (provider=default,fips=yes,input=der,structure=SubjectPublicKeyInfo) { 1.2.840.113549.1.1.1, 2.5.8.1.1, RSA, rsaEncryption } @ default (provider=default,fips=yes,input=der,structure=type-specific) { 1.2.840.113549.1.1.1, 2.5.8.1.1, RSA, rsaEncryption } @ default (provider=default,fips=yes,input=der,structure=rsa) { 1.2.840.113549.1.1.1, 2.5.8.1.1, RSA, rsaEncryption } @ default (provider=default,fips=yes,input=msblob) { 1.2.840.113549.1.1.1, 2.5.8.1.1, RSA, rsaEncryption } @ default (provider=default,fips=yes,input=pvk) { 1.2.840.113549.1.3.1, DH, dhKeyAgreement } @ default (provider=default,fips=yes,input=der,structure=PrivateKeyInfo) { 1.2.840.113549.1.3.1, DH, dhKeyAgreement } @ default (provider=default,fips=yes,input=der,structure=SubjectPublicKeyInfo) { 1.2.840.113549.1.3.1, DH, dhKeyAgreement } @ default (provider=default,fips=yes,input=der,structure=type-specific) { 1.2.840.113549.1.3.1, DH, dhKeyAgreement } @ default (provider=default,fips=yes,input=der,structure=dh) { 1.2.840.10040.4.1, 1.2.840.10040.4.3, 1.3.14.3.2.12, 1.3.14.3.2.13, 1.3.14.3.2.27, DSA, DSA-old, DSA-SHA, DSA-SHA1, DSA-SHA1-old, dsaEncryption, dsaEncryption-old, dsaWithSHA, dsaWithSHA1, dsaWithSHA1-old } @ default (provider=default,fips=yes,input=der,structure=PrivateKeyInfo) { 1.2.840.10040.4.1, 1.2.840.10040.4.3, 1.3.14.3.2.12, 1.3.14.3.2.13, 1.3.14.3.2.27, DSA, DSA-old, DSA-SHA, DSA-SHA1, DSA-SHA1-old, dsaEncryption, dsaEncryption-old, dsaWithSHA, dsaWithSHA1, dsaWithSHA1-old } @ default (provider=default,fips=yes,input=der,structure=SubjectPublicKeyInfo) { 1.2.840.10040.4.1, 1.2.840.10040.4.3, 1.3.14.3.2.12, 1.3.14.3.2.13, 1.3.14.3.2.27, DSA, DSA-old, DSA-SHA, DSA-SHA1, DSA-SHA1-old, dsaEncryption, dsaEncryption-old, dsaWithSHA, dsaWithSHA1, dsaWithSHA1-old } @ default (provider=default,fips=yes,input=der,structure=type-specific) { 1.2.840.10040.4.1, 1.2.840.10040.4.3, 1.3.14.3.2.12, 1.3.14.3.2.13, 1.3.14.3.2.27, DSA, DSA-old, DSA-SHA, DSA-SHA1, DSA-SHA1-old, dsaEncryption, dsaEncryption-old, dsaWithSHA, dsaWithSHA1, dsaWithSHA1-old } @ default (provider=default,fips=yes,input=der,structure=dsa) { 1.2.840.10040.4.1, 1.2.840.10040.4.3, 1.3.14.3.2.12, 1.3.14.3.2.13, 1.3.14.3.2.27, DSA, DSA-old, DSA-SHA, DSA-SHA1, DSA-SHA1-old, dsaEncryption, dsaEncryption-old, dsaWithSHA, dsaWithSHA1, dsaWithSHA1-old } @ default (provider=default,fips=yes,input=msblob) { 1.2.840.10040.4.1, 1.2.840.10040.4.3, 1.3.14.3.2.12, 1.3.14.3.2.13, 1.3.14.3.2.27, DSA, DSA-old, DSA-SHA, DSA-SHA1, DSA-SHA1-old, dsaEncryption, dsaEncryption-old, dsaWithSHA, dsaWithSHA1, dsaWithSHA1-old } @ default (provider=default,fips=yes,input=pvk) { 1.2.840.10045.2.1, EC, id-ecPublicKey } @ default (provider=default,fips=yes,input=der,structure=PrivateKeyInfo) { 1.2.840.10045.2.1, EC, id-ecPublicKey } @ default (provider=default,fips=yes,input=der,structure=SubjectPublicKeyInfo) { 1.2.840.10045.2.1, EC, id-ecPublicKey } @ default (provider=default,fips=yes,input=der,structure=type-specific) { 1.2.840.10045.2.1, EC, id-ecPublicKey } @ default (provider=default,fips=yes,input=der,structure=ec) { 1.2.840.113549.1.1.10, RSA-PSS, RSASSA-PSS, rsassaPss } @ default (provider=default,fips=yes,input=der,structure=PrivateKeyInfo) { 1.2.840.113549.1.1.10, RSA-PSS, RSASSA-PSS, rsassaPss } @ default (provider=default,fips=yes,input=der,structure=SubjectPublicKeyInfo) { 1.2.840.10046.2.1, dhpublicnumber, DHX, X9.42 DH } @ default (provider=default,fips=yes,input=der,structure=PrivateKeyInfo) { 1.2.840.10046.2.1, dhpublicnumber, DHX, X9.42 DH } @ default (provider=default,fips=yes,input=der,structure=SubjectPublicKeyInfo) { 1.2.840.10046.2.1, dhpublicnumber, DHX, X9.42 DH } @ default (provider=default,fips=yes,input=der,structure=type-specific) { 1.2.840.10046.2.1, dhpublicnumber, DHX, X9.42 DH } @ default (provider=default,fips=yes,input=der,structure=dhx) { 1.3.101.110, X25519 } @ default (provider=default,fips=yes,input=der,structure=PrivateKeyInfo) { 1.3.101.110, X25519 } @ default (provider=default,fips=yes,input=der,structure=SubjectPublicKeyInfo) { 1.3.101.111, X448 } @ default (provider=default,fips=yes,input=der,structure=PrivateKeyInfo) { 1.3.101.111, X448 } @ default (provider=default,fips=yes,input=der,structure=SubjectPublicKeyInfo) { 1.3.101.112, ED25519 } @ default (provider=default,fips=yes,input=der,structure=PrivateKeyInfo) { 1.3.101.112, ED25519 } @ default (provider=default,fips=yes,input=der,structure=SubjectPublicKeyInfo) { 1.3.101.113, ED448 } @ default (provider=default,fips=yes,input=der,structure=PrivateKeyInfo) { 1.3.101.113, ED448 } @ default (provider=default,fips=yes,input=der,structure=SubjectPublicKeyInfo) { 1.2.156.10197.1.301, SM2 } @ default (provider=default,fips=no,input=der,structure=PrivateKeyInfo) { 1.2.156.10197.1.301, SM2 } @ default (provider=default,fips=no,input=der,structure=SubjectPublicKeyInfo) { 1.2.156.10197.1.301, SM2 } @ default (provider=default,fips=no,input=der,structure=type-specific) ML-DSA-65 @ default (provider=default,fips=no,input=der,structure=PrivateKeyInfo) ML-DSA-65 @ default (provider=default,fips=no,input=der,structure=SubjectPublicKeyInfo) DER @ default (provider=default,fips=yes,input=der,structure=SubjectPublicKeyInfo) DER @ default (provider=default,fips=yes,input=pem) DER @ default (provider=default,fips=yes,input=der,structure=EncryptedPrivateKeyInfo) root@gcl:/opt/sm2tongsuo/bin#
> > 
> > 
> > Thanks for the feedback. This seems like an environment-specific issue, as I cannot reproduce it locally. Could you please share some details about your build environment so I can try to replicate it? Specifically:
> > 
> > 1. OS and architecture.
> > 2. The commit hash.
> > 3. The full ./config command you used.
> > 4. The output of gcc --version.
> > 
> > Thanks for your help!
> 
> 签发证书还需要在 obj_xref.txt 文件里加ML_DSA_65，文件内容 [obj_xref.txt](https://github.com/user-attachments/files/21545584/obj_xref.txt) 21 ED448 undef ED448 22 #ML_DSA_44 undef ML_DSA_44 23 ML_DSA_65 undef ML_DSA_65 24 #ML_DSA_87 undef ML_DSA_87

Thanks for the feedback! This issue has been fixed.

HOCOOH commented:
> Hi, @HOCOOH, We checked the interoperability of your code with OpenSSL 3.5.2 using the API and command line for signature verification and signing certificate , and found no problems.

Hi, @Maxlsc, thank you for testing this! It's great to hear that no issues were found!

pr000000f reviewed (approved):
This implementation is based on the ML-DSA reference implementation by the Dilithium team and is interoperable with OpenSSL 3.5.0's implementation. I approve this merge.
